editor/tree_search.cpp

```diff
@@ -0,0 +1,654 @@
+
+	// Positioning and sizing
+	set_anchors_and_offsets_preset(LayoutPreset::PRESET_BOTTOM_WIDE);
+	set_v_size_flags(SIZE_SHRINK_CENTER); // Do not expand vertically
+
+	line_edit_search->set_h_size_flags(SIZE_EXPAND_FILL);
+
+	_add_spacer(0.1); // -> Otherwise the lineedits expand margin touches the left border.
+	add_child(line_edit_search);
+	add_child(find_prev_button);
+	add_child(find_next_button);
+	_add_spacer(0.25);
+
+	add_child(check_button_filter_highlight);
+	add_child(label_filter);
+
+	_add_spacer(0.25);
+	add_child(close_button);
+}
+
+void TreeSearchPanel::_add_spacer(float p_width_multiplier) {
+	Control *spacer = memnew(Control);
+	spacer->set_custom_minimum_size(Vector2(8.0 * EDSCALE * p_width_multiplier, 0.0));
+	add_child(spacer);
+}
+
+void TreeSearchPanel::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			// Close callbacks
+			close_button->connect(LW_NAME(pressed), Callable(this, LW_NAME(set_visible)).bind(false));
+			close_button->connect(LW_NAME(pressed), Callable(this, LW_NAME(emit_signal)).bind(LW_NAME(Close)));
+			close_button->set_shortcut(LW_GET_SHORTCUT("limbo_ai/hide_tree_search")); // TODO: use internal shortcut. also sets tooltip...
+			// Search callbacks
+			Callable c_update_requested = Callable(this, LW_NAME(emit_signal)).bind("update_requested");
+			Callable c_text_submitted = Callable(this, LW_NAME(emit_signal)).bind(LW_NAME(text_submitted));
+			Callable c_select_previous_match = Callable(this, LW_NAME(emit_signal)).bind("select_previous_match");
+			find_next_button->connect(LW_NAME(pressed), c_text_submitted);
+			find_prev_button->connect(LW_NAME(pressed), c_select_previous_match);
+
+			line_edit_search->connect(LW_NAME(text_changed), c_update_requested.unbind(1));
+			check_button_filter_highlight->connect(LW_NAME(pressed), c_update_requested);
+			line_edit_search->connect(LW_NAME(text_submitted), c_text_submitted.unbind(1));
+			break;
+		}
+		case NOTIFICATION_THEME_CHANGED: {
+			BUTTON_SET_ICON(close_button, get_theme_icon(LW_NAME(Close), LW_NAME(EditorIcons)));
+			BUTTON_SET_ICON(find_prev_button, get_theme_icon("MoveUp", LW_NAME(EditorIcons)));
+			BUTTON_SET_ICON(find_next_button, get_theme_icon("MoveDown", LW_NAME(EditorIcons)));
+			label_filter->set_text(TTR("Filter"));
+			break;
+		}
+	}
+}
+
+void TreeSearchPanel::_bind_methods() {
+	ADD_SIGNAL(MethodInfo("update_requested"));
+	ADD_SIGNAL(MethodInfo(LW_NAME(text_submitted)));
+	ADD_SIGNAL(MethodInfo("select_previous_match"));
+	ADD_SIGNAL(MethodInfo(LW_NAME(Close)));
+}
+
+TreeSearchPanel::TreeSearchPanel() {
+	_initialize_controls();
+	set_visible(false);
+}
+
+TreeSearch::TreeSearchMode TreeSearchPanel::get_search_mode() const {
+	if (!check_button_filter_highlight || !check_button_filter_highlight->is_pressed()) {
+		return TreeSearch::TreeSearchMode::HIGHLIGHT;
+	}
+	return TreeSearch::TreeSearchMode::FILTER;
+}
+
+String TreeSearchPanel::get_text() const {
+	if (!line_edit_search) {
```

Comment:
`line_edit_search` is allocated in the constructor, so this check is unnecessary

Reply:
Removed.